### PR TITLE
--infra-name command line argument

### LINKED ISSUE #10794: RFE: pod infra container name customization
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

I would like to be able to set name for infra container created inside pod, Additional option in `podman pod create` in form of `--infra-name=name` should be ok. Reasoning: we are using strict naming convention for pods/containers and only infra containers require additional code path to handle its naming divergence.

### PR DESCRIPTION
**Pull request by jvanz**
Adds the new --infra-name command-line argument allowing users to define the name of the infra container

Closes #10794

### CODE REVIEW

File: libpod/options.go [hunk 1]
```diff
@@ -1786,6 +1799,19 @@ func WithInfraCommand(cmd []string) PodCreateOption {
 	}
 }
 
+// WithInfraName sets the infra container name for libpod.
```
Comment (mheon) [inline libpod/options.go:1802]: For a single pod, not libpod

File: libpod/runtime_pod_infra_linux.go [hunk 2]
```diff
@@ -189,7 +189,12 @@ func (r *Runtime) makeInfraContainer(ctx context.Context, p *Pod, imgName, rawIm
 		g.AddLinuxSysctl(sysctlKey, sysctlVal)
 	}
 
-	containerName := p.ID()[:IDTruncLength] + "-infra"
+	containerName := p.config.InfraContainer.InfraName
+	if containerName == "" {
+		containerName = p.ID()[:IDTruncLength] + "-infra"
+	}
+	logrus.Debugf("Infra container name argument: %s", p.config.InfraContainer.InfraName)
```
Comment (mheon) [inline libpod/runtime_pod_infra_linux.go:196]: Can probably drop this one, and just leave the infra name debug below

File: libpod/runtime_pod_infra_linux.go [hunk 3]
```diff
@@ -189,7 +189,12 @@ func (r *Runtime) makeInfraContainer(ctx context.Context, p *Pod, imgName, rawIm
...
+	logrus.Debugf("Infra container name %s", containerName)
```
Comment (mheon) [inline libpod/runtime_pod_infra_linux.go:197]: Upgrade this one to an Infof please

**Comment by mheon** [comment]
/approve
Couple nits, but looks solid overall

File: libpod/runtime_pod_infra_linux.go (hunk 3 above)
Comment (jvanz) [reply libpod/runtime_pod_infra_linux.go:197]: Ops... I forgot to remove these lines. Do you want to keep it as `Infof` ?

Comment (jvanz) [reply libpod/runtime_pod_infra_linux.go:197]: Oh... never mind. I missed the previous comment.

File: cmd/podman/pods/create.go [hunk 4]
```diff
@@ -144,6 +148,10 @@ func create(cmd *cobra.Command, args []string) error {
 			return errors.New("cannot set infra-image without an infra container")
 		}
 		createOptions.InfraImage = ""
+		if cmd.Flag("infra-name").Changed {
```
Comment (rhatdan) [inline cmd/podman/pods/create.go:151]: You don't need any of this.

Just check if createOptions.InfraName != ""

File: cmd/podman/pods/create.go [hunk 5]
```diff
@@ -165,6 +173,12 @@ func create(cmd *cobra.Command, args []string) error {
 				return err
 			}
 		}
+		if cmd.Flag("infra-name").Changed {
```
Comment (rhatdan) [inline cmd/podman/pods/create.go:176]: No need for this, see above.

File: test/system/200-pod.bats [hunk 6]
```diff
@@ -237,6 +239,9 @@ EOF
     # confirm that entrypoint is what we set
     run_podman container inspect --format '{{.Config.Entrypoint}}' $infra_cid
     is "$output" "$infra_command" "infra-command took effect"
+    # confirm that infra container name is set
+    run_podman container inspect --format '{{.Name}}' $infra_cid
+    is "$output" "$infra_name" "infra-name took effect"
```
Comment (rhatdan) [inline test/system/200-pod.bats:244]: Could you verify an error if I set an infra_name to an existing container/pod and make sure it returns a nice error.

**Review by rhatdan — APPROVED** [review]
Please squash your commits into one PR>

**Comment by rhatdan** [comment]
We have renamed the master branch to the main branch,   You need to rebase to the latest code base.

**Comment by mheon** [comment]
DCO validation is presently complaining that your commits do not have a valid signoff - you can add one with `git commit --amend -s` (easiest way is probably to squash the 4 commits into one, and then add a signoff)

**Comment by mheon** [comment]
Ah, manpage checker is noting that the new flag has not been added to docs/source/markdown/podman-pod-create.1.md

**Comment by mheon** [comment]
Test failures looked like infra, restarted

**Comment by mheon** [comment]
Tests green

**Comment by mheon** [comment]
LGTM
@containers/podman-maintainers PTAL

File: test/system/200-pod.bats [hunk 7]
```diff
@@ -205,6 +205,7 @@ function random_ip() {
     # entrypoint to confirm that --infra-command will override.
     local infra_image="infra_$(random_string 10 | tr A-Z a-z)"
     local infra_command="/pause_$(random_string 10)"
+    local infra_name="testing-infra-container"
```
Comment (edsantiago) [inline test/system/200-pod.bats:208]: I realize I'm late in the game and that tests are green. If you have to resubmit for some reason, please fix this so `infra_name` includes a random element. Randomization is critical in testing, it helps catch all manner of subtle bugs.

File: docs/source/markdown/podman-pod-create.1.md [hunk 8]
```diff
@@ -75,6 +75,10 @@ The command that will be run to start the infra container. Default: "/pause".
 
 The image that will be created for the infra container. Default: "k8s.gcr.io/pause:3.1".
 
+#### **--infra-name**=*name*
+
+The name that will be used in the pod's infra container.
```
Comment (edsantiago) [inline docs/source/markdown/podman-pod-create.1.md:80]: Shouldn't this be "used **for**", not "in"?

File: cmd/podman/pods/create.go [hunk 9]
```diff
@@ -144,6 +148,10 @@ func create(cmd *cobra.Command, args []string) error {
 			return errors.New("cannot set infra-image without an infra container")
 		}
 		createOptions.InfraImage = ""
+		if createOptions.InfraName != "" {
+			return errors.New("cannot set infra-name without an infra container")
+		}
+		createOptions.InfraName = ""
```
Comment (edsantiago) [inline cmd/podman/pods/create.go:154]: I'm confused. You just tested against empty string. The only way the code can get here is if `InfraName` is already empty string. What purpose does this line serve?

**Review by edsantiago — COMMENTED** [review]
Suggestions and one question about unnecessary (I think) code

**Review by edsantiago — APPROVED** [review]
My concerns are addressed, thank you. I don't see an acknowledgment of @rhatdan's concern from last week, re: emitting a helpful message if `infra-name` is in use. In particular, the current error message has what I believe is an unnecessary extra message:
```console
$ ./bin/podman pod create --name=mypod --infra-name=foo
ERRO[0000] Error freeing pod lock after failed creation: no such file or directory
Error: error adding Infra Container: error creating container storage: the container name "foo" is already in use by "ba1380cfd2c6a3e0a7d03b143c1428661434efe073f99747b0bb25bce5f5d156". You have to remove that container to be able to reuse that name.: that name is already in use
```
(I see no reason for a human user to see that "error freeing pod lock" message)

File: test/system/200-pod.bats (hunk 6 above)
Comment (jvanz) [reply test/system/200-pod.bats:244]: I'll fix this

Comment (jvanz) [reply test/system/200-pod.bats:244]: The current error is: 
```
jvanz@earth:~/repositories/podman> ./bin/podman  pod create --infra-name testing-infra-name 
ERRO[0000] Error freeing pod lock after failed creation: no such file or directory 
Error: error adding Infra Container: error creating container storage: the container name "testing-infra-name" is already in use by "0b25cc744fda353093970052ee80f5dd5c43917590854615ad59dcf5f57f7984". You have to remove that container to be able to reuse that name.: that name is already in use
```

It seems reasonable to me. What do you think? 
I will add a test to validate the expected behaviour.

Comment (jvanz) [reply test/system/200-pod.bats:244]: Test added.

**Comment by jvanz** [comment]
As far as I can see, the tests failed due time out error. Can we rerun them?

**Comment by mheon** [comment]
All system tests are failing, and those don't look like timeouts

**Comment by jvanz** [comment]
> All system tests are failing, and those don't look like timeouts

Ops, my bad. There is typo in my last change without testing before push... =\

**Comment by jvanz** [comment]
I just rebased the commit on top of the latest main branch. It seems Cirrus CI is having problems running the jobs.

**Comment by jvanz** [comment]
:rocket:

**Comment by rhatdan** [comment]
/lgtm
Thanks @jvanz

**Pull request merged**